doc/design/parallel-node-config.md

```diff
@@ -0,0 +1,214 @@
+
+1. config daemon mark the node as `Drain_Required` or `Reboot_Required`
+2. operator drain controller reconcile loop find the node SriovNetworkPoolConfig
+   1. if number of `Draining` nodes is great or equal to the `MaxUnavailable` the operator will re-queue the request
+   2. if number of `Draining` nodes is lower than the `MaxUnavailable` the operator will start the draining process
+5. on Openshift platform we will pause the machine config pool related to the node
+6. the operator will start the drain process
+   1. if `Drain_Required` the operator will remove ONLY pods used sriov devices
+   2. if `Reboot_Required` the operator will remove ALL the pods on the system
+9. operator moves the `sriovnetwork.openshift.io/current-state` annotation to `DrainComplete`
+10. daemon will continue to the configuration when it's done it will move back the `sriovnetwork.openshift.io/state` annotation to `Idle`
+11. operator runs the complete drain to remove the cordon and mark the `sriovnetwork.openshift.io/current-state` annotation to `Idle`
+
+### API Extensions
+
+#### Extend existing CR SriovNetworkPoolConfig
+SriovNetworkPoolConfig is used only for OpenShift to provide configuration for
+OVS Hardware Offloading. We can extend it to add configuration for the drain
+pool. E.g.:
+
+```golang
+// SriovNetworkPoolConfigSpec defines the desired state of SriovNetworkPoolConfig
+type SriovNetworkPoolConfigSpec struct {
+    ...
+	
+    // nodeSelector specifies a label selector for Nodes
+    NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+    // maxUnavailable defines either an integer number or percentage
+    // of nodes in the pool that can go Unavailable during an update.
+    //
+    // A value larger than 1 will mean multiple nodes going unavailable during
+    // the update, which may affect your workload stress on the remaining nodes.
+    // Drain will respect Pod Disruption Budgets (PDBs) such as etcd quorum guards,
+    // even if maxUnavailable is greater than one.
+    MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
+}
+```
+
+```yaml
+apiVersion: v1
+kind: SriovNetworkPoolConfig
+metadata:
+  name: pool-1
+  namespace: network-operator
+spec:
+  maxUnavailable: "20%"
+  nodeSelectorTerms:
+    - matchExpressions:
+      - key: some-label
+        operator: In
+        values:
+          - val-2
+    - matchExpressions:
+      - key: other-label
+        operator: "Exists"
+```
+
+Default pool should be always created with the lowest priority:
+```yaml
+apiVersion: v1
+kind: SriovNetworkPoolConfig
+metadata:
+  name: default
+  namespace: network-operator
+spec:
+  maxUnavailable: 1
+```
+
+Once this change will be implemented `SriovNetworkPoolConfig` configuration will be applied both to vanilla Kubernetes
+and OpenShift clusters.
+
+### Implementation Constraints
+
+Node can only be part of one pool beside the default pool that catch all the nodes.
+
+### Upgrade & Downgrade considerations
```

Comment:
do we want to cover upgrade / downgrade in the middle of node sriov config ?

e.g config daemon started cordon drain and we upgrade the operator ?

i guess we should say "dont do that"

Reply:
it's not possible to say don't do it because there are systems that as part of the upgrade for example OCP we reboot the nodes to install a new version of the operating system so you can have a state where one of the daemons is doing a configuration

Reply:
ack, so need to add some info about it IMO.

Reply:
do we want in this PR or the next one that will address the transition between the two annotations?

Reply:
controller behaviour WRT handling such case should be mentioned in this PR.
also we may have +1 node being cordoned in this case IMO, we should mention this as well.

moving to the use of a single object (nodestate) is a different topic (and different PR).

Reply:
the reason we are doing this in two steps is to not have this issue.
meaning that in the future when we drop the node label if it's in an upgrade case the existing operator will already have the code in place to add both annotations in parallel so you can't have a problem that some daemons use the old node annotation and some use the new annotation.

Reply:
in our current code, the config daemon will drain if no other node has `sriovnetwork.openshift.io/state: Draining` annotation.

i went over #555 and we never set "Draining" on the node object. so basically i dont understand why we need to set the node annotations (to a value which old config daemon is not aware of).

see: https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/82a6d6fdce71bd88a0d9368fb1750488e9a8e4e2/pkg/daemon/daemon.go#L381C49-L381C73

which is why i wanted to have more details about it.